### ISSUE TITLE
How to use DALI in mxnet？

### ISSUE THREAD

**Issue by pawopawo** [NVIDIA/DALI#1270]
I saw how to process ImageNet data with dali as `dali_train_iter`, `dali_val_iter` at https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/mxnet/mxnet-resnet50.html
But did not see how to extract the data in `dali_train_iter`, `dali_val_iter`.

**Comment by awolant** [comment]
Hi, thanks for the question.
Could you be more specific on what you mean by extracting the data?
[This example](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/mxnet/mxnet-various-readers.html) shows hot to obtain the pipeline outputs form the MXNet iterator (last code cell). Did you mean something like this?

**Comment by xxsgcjwddsg** [comment]
If on a single machine with 8 gpus, the method for pytorch to extract data is：
```
for step, data in enumerate(loader):
         Input = torch.cat([data[j]["data"] for j in range(args.num_shards)], dim=0)
         Target = torch.cat([data[j]["label"] for j in range(args.num_shards)], dim=0).squeeze().long()
```, but I don't see how to use the for loop to extract data in https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/mxnet/mxnet-resnet50.html.

In the [new example](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/mxnet/mxnet-various-readers.html) you provided, I saw：    
```
for i, data in enumerate(dali_iter):
        if i >= ITERATIONS:
            break
        # Testing correctness of labels
        for d in data:
            label = d.label[0].asnumpy()
            image = d.data[0]
```
But if use a single machine with 8 gpus, how should the code be written, need concat like pytorch?

**Comment by awolant** [comment]
To clarify, for multi GPU training  in PyTorch, we do not concatenate outputs. In this case there are multiple processes with separate pipelines processing different samples.

Going back to MXNet, if you are looking for more complete example you can check out [Deep Learning Examples MXNet ResNet](https://github.com/NVIDIA/DeepLearningExamples/tree/master/MxNet/Classification/RN50v1.5). It's ready to use and can handle multiple GPUs. Also some links from #1228 might be useful for you.

**Comment by xxsgcjwddsg** [comment]
Thank you very much!
The code I am training on imagenet comes from [train_imagenet](https://gluon-cv.mxnet.io/_downloads/3bb06a6d6d085b1bb501b30aaf6c21c5/train_imagenet.py). I want to replace the data read function with DALI. Do you have any suggestions for replacement?

In [Gluon example](https://docs.nvidia.com/deeplearning/sdk/dali-developer-guide/docs/examples/mxnet/gluon.html),I saw using：
```
for batches in dali_iter: # Using DALI iterator
         ##################################################
         # (1) Update D network: maximize log(D(x)) + log(1 - D(G(z)))
         ##########################
         Data = batches[0].data[0] # extracting the batch for device 0
```
read a batch of data on device 0. If there are 8 gpus, means have 8 devices? How should I change the code? In addition, how should the label be read?

**Comment by awolant** [comment]
@Kh4L is the most knowledgeable about this code

**Comment by Kh4L** [comment]
Hi @xxsgcjwddsg ,

`batches` is a list of per GPU MXNet `DataBatch` (more about it [here](https://mxnet.incubator.apache.org/api/python/io/io.html)).
Consequently, if you provide 8 Pipelines to the iterator, `batches` will be a list of 8 MXNet `DataBatch`.

This comes from how MXNet symbolic API works.

When it comes to Gluon, your model will expects a list (of size `num_gpus`) of NDArrays for each input. Let's say your DALI pipeline outputs 2: one data and one label. In Gluon you will have to write this:
```
for batch in dali_iter:
    data = [d.data[0] for d in batch]
    labels = [d.label[0] for d in batch]
    l = loss(model(data), labels)
```
`data` and `labels` will be two lists of 8 NDArrays.

You can take a look at those lines, in a real-world example:
https://github.com/dmlc/gluon-cv/blob/6a0e4d0513887b77efe3361738e83145a02e7f9a/scripts/detection/ssd/train_ssd.py#L306

**Comment by xxsgcjwddsg** [comment]
Thanks！
```
for batch in dali_iter:
    data = [d.data[0] for d in batch]
    labels = [d.label[0] for d in batch]
    l = loss(model(data), labels)
```
Is the processing of the above code similar to `batch_fn` in the `https://github.com/dmlc/gluon-cv/blob/6a0e4d0513887b77efe3361738e83145a02e7f9a/scripts/classification/imagenet/train_imagenet.py#L198`?

My goal is to replace `[get_data_rec](https://github.com/dmlc/gluon-cv/blob/6a0e4d0513887b77efe3361738e83145a02e7f9a/scripts/classification/imagenet/train_imagenet.py#L185)` with Dali functions.

**Comment by Kh4L** [comment]
> Is the processing of the above code similar to `batch_fn` in the https://github.com/dmlc/gluon-cv/blob/6a0e4d0513887b77efe3361738e83145a02e7f9a/scripts/classification/imagenet/train_imagenet.py#L198?

Yes it produces the same output.

What is your model and what is the training data?

**Comment by xxsgcjwddsg** [comment]
Thanks! I'm sorry to reply so late.
The model I trained is shufflenetv2. The dataset I used is imagent2012 and the format is rec. I also encountered a problem in reading rec data with DALI, which I raised under another [issue](https://github.com/NVIDIA/DALI/issues/1279).

**Comment by Kh4L** [comment]
@xxsgcjwddsg the error telling you that the second input of the loss function is on the CPU (`cpu(0)` context) while it is expected to be on the GPU (`gpu(0)` context).

Do you transfer the label on the GPU in your DALI pipeline?
You should have:
```
label = label.gpu()
# or simply
return image, label.gpu()
```
in the `define_graph`.
